src/lib/SoftHSM.cpp

```diff
@@ -11383,23 +11524,76 @@ CK_RV SoftHSM::deriveSymmetric
 			bOK = bOK && osobject->setAttribute(CKA_LOCAL,false);
 
 			// Common Secret Key Attributes
-			if (baseKey->getBooleanValue(CKA_ALWAYS_SENSITIVE, false))
-			{
-				bool bAlwaysSensitive = osobject->getBooleanValue(CKA_SENSITIVE, false);
-				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE,bAlwaysSensitive);
-			}
-			else
-			{
-				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE,false);
-			}
-			if (baseKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, true))
-			{
-				bool bNeverExtractable = osobject->getBooleanValue(CKA_EXTRACTABLE, false) == false;
-				bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE,bNeverExtractable);
-			}
-			else
-			{
-				bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE,false);
+			if (pMechanism->mechanism == CKM_CONCATENATE_BASE_AND_KEY) {
+				// [PKCS#11 v2.40, 2.31.3]
+				// If either of the two original keys has its CKA_SENSITIVE attribute
+				// set to CK_TRUE, so does the derived key.  If not, then the derived
+				// key’s CKA_SENSITIVE attribute is set either from the supplied template
+				// or from a default value.
+				bool bSensitive = baseKey->getBooleanValue(CKA_SENSITIVE, true) ||
+								  otherKey->getBooleanValue(CKA_SENSITIVE, true);
+				if (bSensitive) {
+					bOK = bOK && osobject->setAttribute(CKA_SENSITIVE, true);
+				}
+				// If either of the two original keys has its CKA_EXTRACTABLE attribute
+				// set to CK_FALSE, so does the derived key.  If not, then the derived
+				// key’s CKA_EXTRACTABLE attribute is set either from the supplied template
+				// or from a default value.
+				bool bExtractable = baseKey->getBooleanValue(CKA_EXTRACTABLE, false) ||
+									otherKey->getBooleanValue(CKA_EXTRACTABLE, false);
```

Comment:
The logic here is wrong, IMHO. One of the keys being non-extractable should make the resulting key non-extractable.

```suggestion
				bool bExtractable = baseKey->getBooleanValue(CKA_EXTRACTABLE, false) &&
									otherKey->getBooleanValue(CKA_EXTRACTABLE, false);
```

Reply:
You're correct. I've fixed it and added test cases.